### PR TITLE
Rework ajax errors

### PR DESCRIPTION
**Pull request by naphets123**

### CODE REVIEW

**Comment by HaGuesto** [comment]
[Rework ajax errors on desktop screen #large](https://trello.com/c/JLdkcUCd/133-rework-ajax-errors-on-desktop-screen-large)

File: assets/js/custom.js [hunk 1]
```diff
@@ -151,6 +142,47 @@ function offsetAnchor() {
...
+            }
+            
+
+        if (result.action) {
+            eval(result.action);
+            }
+
+            
+    }
+    else
```
Comment (HaGuesto) [inline assets/js/custom.js]: This if/else does not make sense and causes many tests to fail.
How you have written it it only posts messages if there was no error on the server-side. However, there are cases like login errors were other error messages should be shown.

File: assets/js/magic.js [hunk 2]
```diff
@@ -422,9 +422,10 @@ $(function() {
                         dataType: "json",
                         success: function(checkresult) {
                             if (checkresult.success) {
-                                AjaxFormSubmit(form);
```
Comment (HaGuesto) [inline assets/js/magic.js]: @naphets123 Why did you move this line?

Comment (naphets123) [reply assets/js/magic.js]: i put it back

File: assets/js/custom.js (hunk 1 above)
Comment (naphets123) [reply assets/js/custom.js]: Ok, I removed the whole conditional on succes, it was actually just used for everything except message/redirect/action

Comment (HaGuesto) [reply assets/js/custom.js]: However, you removed to much now. 
You neglected the case 3) now from the trello card.
>Distinguish between three errors:
>1) Ajax request does not get a response from the server. The notification should be
>"Cannot connect to Boxwise - please check your Internet connection."
>2) The response is a json looking like {'success': false, 'message':'Something something', ....}
>If a message is specified please show it to the user.
>3) A response is received, but it is not in the format of 2). The notification should be
>"Something went wrong - please inform your coordinator."

Comment (HaGuesto) [reply assets/js/custom.js]: I will tag the line where you should put it.

File: assets/js/custom.js [hunk 3]
```diff
@@ -151,6 +142,38 @@ function offsetAnchor() {
         window.scrollTo(window.scrollX, window.scrollY - 100);
     }
 }
+
+function AjaxCheckSuccess(result){
+        if (result.message){
```
Comment (HaGuesto) [inline assets/js/custom.js]: Add here something like
```
if (result.success does not exist) {
            var n = noty({
                text: "Something went wrong - please inform your coordinator.",
                type: "error"
                });
        }

File: assets/js/custom.js [hunk 4]
```diff
@@ -151,6 +142,43 @@ function offsetAnchor() {
...
+            var n = noty({
+                text: result.message,
+                type: result.success ? "success" : "error"
+                });
+        }
+        if (result.redirect){
```
Comment (HaGuesto) [inline assets/js/custom.js]: This if clause does not make sense. It should look like
```
                if (result.redirect) {
                    if (result.message) {
                        setTimeout(function() {
                            execReload(result.redirect);
                        }, 1500);
                    } else {
                        execReload(result.redirect);
                    }
                }
```

Comment (HaGuesto) [reply assets/js/custom.js]: you left out the second if clause. The redirect is timed out for the time a notification needs to be shown.

File: assets/js/custom.js [hunk 5]
```diff
@@ -378,19 +388,10 @@ function getSizes() {
...
-            }
+            AjaxCheckSuccess(result);
         },
         error: function(result) {
-            var n = noty({
-                text:
-                    "Something went wrong, maybe the internet connection is a bit choppy",
-                type: "error"
-            });
+            AjaxError(result);
```
Comment (jamescrowley) [inline assets/js/custom.js]: Just a little note that you can simplify

```
error: function(result) {
     AjaxError(result);
},
```

to

`error: AjaxError,`

and so on - ie just pointing to the function definition :)

File: assets/js/custom.js (hunk 4 above)
Comment (naphets123) [reply assets/js/custom.js]: Ok I changed this and applied the shorter syntax james proposed below. I think this fixes the issue you saw with the login tests

File: assets/js/custom.js [hunk 6]
```diff
@@ -378,20 +386,10 @@ function getSizes() {
...
+            AjaxCheckSuccess(result);
         },
-        error: function(result) {
-            var n = noty({
-                text:
-                    "Something went wrong, maybe the internet connection is a bit choppy",
-                type: "error"
-            });
-        }
+        error: AjaxError(result)
```
Comment (HaGuesto) [inline assets/js/custom.js]: This throws a timeout error in Test 3_2.

The browser console shows a Reference Error.
```
custom.js?v=7:393 Uncaught ReferenceError: result is not defined
    at getSizes (custom.js?v=7:393)
    at HTMLSelectElement.onchange (?action=stock_edit&origin=stock:201)
    at Object.trigger (minified.js:1)
    at HTMLSelectElement.<anonymous> (minified.js:1)
    at Function.each (minified.js:1)
    at r.fn.init.each (minified.js:1)
    at r.fn.init.trigger (minified.js:1)
    at d.triggerChange (minified.js:348)
    at d.onSelect (minified.js:348)
    at d.selectHighlighted (minified.js:348)
```

Comment (HaGuesto) [reply assets/js/custom.js]: @naphets123 I figured the Error out and fixed it. Ready to merge now.
